tests/MongoDB.Driver.Tests/AsyncCursorTests.cs

```diff
@@ -78,17 +80,69 @@ public void KillCursor_should_actually_work()
                 cursor = collection.FindSync("{}");
                 cursor.MoveNext();
 
-                var cursorId = ((AsyncCursor<BsonDocument>) cursor)._cursorId();
+                var cursorId = ((AsyncCursor<BsonDocument>)cursor)._cursorId();
                 cursorId.Should().NotBe(0);
                 cursor.Dispose();
 
                 var desiredResult = BsonDocument.Parse($"{{ \"cursorsKilled\" : [{cursorId}], \"cursorsNotFound\" : [], " +
                     $"\"cursorsAlive\" : [], \"cursorsUnknown\" : [], \"ok\" : 1.0 }}");
-                var result = ((CommandSucceededEvent) eventCapturer.Events[0]).Reply;
+                var result = ((CommandSucceededEvent)eventCapturer.Events[0]).Reply;
                 result.IsSameOrEqualTo(desiredResult);
             }
         }
 
+        [SkippableFact]
+        public void Tailable_cursor_should_be_able_to_be_cancelled_from_a_different_thread_with_expected_result()
+        {
+            RequireServer.Check().Supports(Feature.TailableCursor);
+
+            string testCollectionName = "test";
+            string testDatabaseName = "test";
+
+            var client = DriverTestConfiguration.Client;
+            var database = client.GetDatabase(testDatabaseName);
+            var collection = database.GetCollection<BsonDocument>(testCollectionName);
+
+            DropCollection(client, testDatabaseName, testCollectionName);
+            var createCollectionOptions = new CreateCollectionOptions()
+            {
+                Capped = true,
+                MaxSize = 1000
+            };
+            database.CreateCollection(testCollectionName, createCollectionOptions);
+            collection.InsertOne(new BsonDocument()); // tailable cursors don't work on an empty collection
+
+            using (var cancellationTokenSource = new CancellationTokenSource())
+            {
+                var findOptions = new FindOptions<BsonDocument>()
+                {
+                    BatchSize = 1,
+                    CursorType = CursorType.TailableAwait
+                };
+                var cursor = collection.FindSync(FilterDefinition<BsonDocument>.Empty, findOptions);
+                var enumerator = cursor.ToEnumerable(cancellationTokenSource.Token).GetEnumerator();
+
+                var semaphore = new SemaphoreSlim(0);
+                var thread = new Thread(() =>
+                {
+                    semaphore.Wait();
+                    cancellationTokenSource.Cancel();
+                });
+                thread.Start();
+
+                var exception = Record.Exception((Action)(() =>
+                {
+                    while (true)
```

Comment:
Does this need to be an infinite loop? This seems to work on my machine:

```
var exception = Record.Exception((Action)(() =>
                {
                    // while (true)
                    {
                        _ = enumerator.MoveNext();
                        semaphore.Release(1);
                        _ = enumerator.MoveNext();
                    }
                }));
```

Reply:
I don't think it has to be an infinite loop, but calling MoveNext only twice seems like a race condition.

There is no guarantee that the call to Release(1) will result in cancellation being requested before the second call to MoveNext (but it might sometimes...).

It's probably right to be worried about the infinite part... if the test fails because cancellation doesn't happen then the test runner will be stuck here until the whole test process times out. Maybe that doesn't matter because that can't happen unless there is a bug.

Reply:
Discussed offline with Robert! My current understanding is as follows:

If we hypothetically had not used a loop, the second call to `moveNext()` would block because there are no more documents in the collection. This call would block until the new thread called the `cancellationTokenSource.Cancel()`. So we now believe there would not be a race condition. No changes should be needed though because the code as committed will still work 😸